### PR TITLE
Python 3.11

### PR DESCRIPTION
**Pull request by Bo98**
I'm opening this to discuss changes that I've proposed to be made to Python 3.11. These don't have to happen if people disagree. I do not plan to backport these.

There are two areas of concern that I'm hoping to address:

1. Parity with how Apple and Python themselves ship Python. Even MacPorts.
2. Legality of the effect of GPL dependencies on dependents, given how widespread our Python is now being used. Notably we can't exactly legally ship anything that might have been using readline or gdbm with something that uses GPL-2.0-only or OpenSSL 1.1.

To achieve this, this PR has the following changes:

* gdbm is now available as a separate formula `python-gdbm@3.11`
  - Compatibility concerns:
    * `dbm.gnu` is not available until you install that separate formula.
    * `dbm.ndbm` no longer uses gdbm_compat. It uses the system ndbm implementation (on macOS) and Berkeley DB 5 on Linux.
      - Users who need to keep a database created by `dbm.ndbm` in an older version of Python may experience compatibility issues. You'll need to read your database using the older version of Python and convert to another format.
    * `dbm` defaulted to `dbm.gnu`, and will still do when `python-gdbm@3.11` is installed.
    * For those who can't migrate yet, Python 3.10 will continue to be available for at least 4 years.
* readline implementation has been switched to libedit
  - Compatibility concerns:
    * API wise, the two are largely compatible with one exception: the initialization file may have a different format, as documented at https://docs.python.org/3/library/readline.html.
    * For the Python Shell/REPL, `.python_history` may have compatibility issues between libedit and readline.
      - This incompatbility already existed if you switched between Homebrew Python and Apple Python.
      - If your previous history file was _never_ written by libedit before (i.e. you have never used another Python shell other than Homebrew) then it may reset upon using Homebrew Python 3.11's shell. Otherwise it will typically convert ok (at least from my limited testing).
      - Spaces in history items will appear "corrupt" when going _back_ to a previous version of Python that still uses GNU readline.
      - We could patch to change the history filename if absolutely necessary (though arguably the previous GNU readline implementation should have done that instead).
  - Note that SQLite, another Python dependency, also uses readline. Ideally that would change too but it isn't actually used in libsqlite3 so can probably get away with it if there's strong opposition to changing that.

We don't have to do these changes, but if we don't we'll need better auditing of where the readline and dbm modules are used and the licensing compatibility.

### CODE REVIEW

**Comment by yurikoles** [comment]
3.11 is [released](https://discuss.python.org/t/python-3-11-0-final-is-now-available/20291).

**Comment by cclauss** [comment]
Do we _need_ to include `gdbm` and `tkinter` in this pull request?  It can easily take six months to get a new version of Python to be the homebrew default.  The more complex this PR is, the longer it will take for it to be ratified.  Small moves.

**Comment by Bo98** [comment]
> Do we _need_ to include `gdbm` and `tkinter` in this pull request? It can easily take six months to get a new version of Python to be the homebrew default. The more complex this PR is, the longer it will take for it to be ratified. Small moves.

We will not be making breaking changes in a patch release.

tkinter shouldn't even be a problem given we've done the same for 2 years now?

If there's strong opposition to gdbm I can of course drop that. Though no one has voiced either way yet, probably given it was rc2 I tested. Will push the final version in a few minutes.

**Review by iMichka — APPROVED** [review]
Looks good to me.

Just one question, should we add `--disable-dbm` to `python@3.11` to avoid opportunistic building/linking?

**Comment by iMichka** [comment]
Might also want one more review from someone else, in case I missed something during my review.

**Comment by Bo98** [comment]
> Just one question, should we add `--disable-dbm` to `python@3.11` to avoid opportunistic building/linking?

`--with-dbmliborder` controls that. I made sure it does not contain `gdbm`.

File: Formula/berkeley-db@5.rb
Comment (MikeMcQuaid) [inline Formula/berkeley-db@5.rb:7]: Do we definitely need a `revision` here?

File: Formula/python-gdbm@3.11.rb
Comment (MikeMcQuaid) [inline Formula/python-gdbm@3.11.rb:10]: Not seen this DSL before but: I really like it. To attempt a nerd-snipe: if anyone else agrees with me it might be cool to allow a similar DSL for `install` or `post_install` or similar for versioned formulae to share more code. If anyone else agrees: I'll write up an issue. CC @Homebrew/maintainers for thoughts.

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me! All questions are non-blocking except if we can remove the `revision` (if we can't: don't and ship regardless).

File: Formula/berkeley-db@5.rb
Comment (Bo98) [reply Formula/berkeley-db@5.rb:7]: This adds new symbols to the library which Python uses. If anyone has the previous version, Python will error/crash when you use `dbm.ndbm`.

Comment (MikeMcQuaid) [reply Formula/berkeley-db@5.rb:7]: Yes, that seems like a good reason 😅

File: Formula/python-gdbm@3.11.rb
Comment (EricFromCanada) [reply Formula/python-gdbm@3.11.rb:10]: It's [documented](https://docs.brew.sh/Brew-Livecheck#referenced-formulacask) but rarely used for versioned formulae; it's instead found in those that are derived from / entangled with another formula, e.g. [lemon](https://formulae.brew.sh/formula/lemon).

Comment (MikeMcQuaid) [reply Formula/python-gdbm@3.11.rb:10]: @EricFromCanada Yeh, I was more thinking a differing DSL something like

```ruby
def install
  formula "python@3.11"
end
```

or something

**Comment by Bo98** [comment]
No objections after 24 hours of green CI. We're good to go.

**Comment by Bo98** [comment]
> It can easily take six months to get a new version of Python to be the homebrew default.

Just to touch on this. I do not think we should be taking that long. Half of the point of moving to a versioned formulae approach was to detach the default from dependents.

If this is proven stable enough with no issues reported after a few weeks, and we've got basics supporting 3.11 like `six` (i.e. the few formulae which support multiple versions of Python), I don't see why we can't change the default by the end of the year.

If we can't, then I consider the process to have failed somewhere given everything we ship should be able to work regardless of what `python3` points to.

**Comment by cclauss** [comment]
> I don't see why we can't change the default by the end of the year.

This pull request went really smoothly!  Thanks for making this happen.  Brew rules!

**Comment by carlocab** [comment]
I'm unclear as to why we didn't change the default now, actually. Not sure why support in `six`, etc. should be a prerequisite?

**Comment by ahankinson** [comment]
Thank you so much for this, and so quickly! Is there any way to get this new formula to install itself as `python3` on the command-line? I noticed this with `brew info python@3.11`:

```
If you do not need a specific version of Python, and always want Homebrew's `python3` in your PATH:
  brew install python3
```

but this seems to leave the case of needing a specific version of Python (3.11) but also always wanting Homebrew's `python3` on my PATH to point to that specific version. Otherwise `python3` points to the Apple-installed version, which is definitely not what is wanted. 

In the past `brew link python@3.11` worked to "install" a specific version of Python as `python3`, but it seems like this no longer works...

**Comment by carlocab** [comment]
> We could definitely try change the default now though. Open a separate PR so we can do a CI run to check.

Yep, we can do that when we catch up with the PR backlog.

**Comment by cclauss** [comment]
@ahankinson You can `brew install python-launcher` and then use `py` instead of `python3`.

`py --list` to see the versions of Python installed.

**Pull request closed without merging**
